lib/host/wasi/inode-win.cpp

```diff
@@ -687,18 +2146,102 @@ WasiExpect<void> INode::sockGetPeerAddr(uint8_t *, uint32_t *,
 }
 
 __wasi_filetype_t INode::unsafeFiletype() const noexcept {
+
+  // TODO: Find equivalents to the other file types
+  // To be completed along with other similar functions
+
+  if (unlikely(GetFileInformationByHandle(Handle, &(*FileInfo))) == FALSE) {
+    return __WASI_FILETYPE_UNKNOWN;
+  }
+
+  switch ((*FileInfo).dwFileAttributes) {
+  case FILE_ATTRIBUTE_DIRECTORY:
+    return __WASI_FILETYPE_DIRECTORY;
+  case FILE_ATTRIBUTE_NORMAL:
+    return __WASI_FILETYPE_REGULAR_FILE;
+  case FILE_ATTRIBUTE_REPARSE_POINT:
+    return __WASI_FILETYPE_SYMBOLIC_LINK;
+  }
+
+  switch (GetFileType(Handle)) {
+  case FILE_TYPE_CHAR:
+    return __WASI_FILETYPE_CHARACTER_DEVICE;
+  }
+  return __WASI_FILETYPE_UNKNOWN;
+
   return __WASI_FILETYPE_UNKNOWN;
 }
 
 WasiExpect<__wasi_filetype_t> INode::filetype() const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+
+  // TODO: Find equivalents to the other file types
+  // To be completed along with other similar functions
+
+  if (unlikely(GetFileInformationByHandle(Handle, &(*FileInfo))) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  switch ((*FileInfo).dwFileAttributes) {
+  case FILE_ATTRIBUTE_DIRECTORY:
+    return __WASI_FILETYPE_DIRECTORY;
+  case FILE_ATTRIBUTE_NORMAL:
+    return __WASI_FILETYPE_REGULAR_FILE;
+  case FILE_ATTRIBUTE_REPARSE_POINT:
+    return __WASI_FILETYPE_SYMBOLIC_LINK;
+  }
+
+  switch (GetFileType(Handle)) {
+  case FILE_TYPE_CHAR:
+    return __WASI_FILETYPE_CHARACTER_DEVICE;
+  }
+  return __WASI_FILETYPE_UNKNOWN;
 }
 
-bool INode::isDirectory() const noexcept { return false; }
+bool INode::isDirectory() const noexcept {
+  if (FileInfo.has_value()) {
```

Comment:
Looks like this section can be a function.
`isSymlink` has the same pattern.
```
if (!FileInfo.has_value()) {}
plus
  if (Handle != INVALID_HANDLE_VALUE &&
      unlikely(GetFileInformationByHandle(Handle, &(*FileInfo))) == FALSE) {
    return false;
  }
```

This pattern is what we do in the `inode-linux.cpp: updateStat()`

Reply:
But how to generalize the `if (!FileInfo.has_value()) {...}` part? The other functions have different return values.

Reply:
I have defined a new function to update the `FileInfo` for this purpose.